File: src/termometro/scorer.py
from __future__ import annotations

import unicodedata
from collections import Counter
from enum import StrEnum


class Tile(StrEnum):
    GREEN = "🟩"
    YELLOW = "🟨"
    BLACK = "⬛"
    SPACE = "  "  # two ASCII spaces, matching one wide emoji column


GREEN = Tile.GREEN
YELLOW = Tile.YELLOW
BLACK = Tile.BLACK
SPACE = Tile.SPACE


def _strip_accents(text: str) -> str:
    decomposed = unicodedata.normalize("NFD", text)
    return "".join(ch for ch in decomposed if unicodedata.category(ch) != "Mn")


def _normalize(text: str, *, case_insensitive: bool, ignore_accents: bool) -> str:
    if ignore_accents:
        text = _strip_accents(text)
    if case_insensitive:
        text = text.casefold()
    return text


def score(
    guess: str,
    answer: str,
    *,
    case_insensitive: bool = True,
    ignore_accents: bool = True,
) -> list[Tile]:
    """Return per-character Wordle tiles comparing guess against answer.

    Uses the classic two-pass algorithm so duplicate letters in the guess do
    not double-count occurrences in the answer: greens consume their letter
    first, then yellows are assigned only while the answer still has spare
    matches.
    """
    if len(guess) != len(answer):
        raise ValueError(
            f"guess and answer must have the same length (got {len(guess)} vs {len(answer)})"
        )

    normalized_guess = _normalize(
        guess, case_insensitive=case_insensitive, ignore_accents=ignore_accents
    )
    normalized_answer = _normalize(
        answer, case_insensitive=case_insensitive, ignore_accents=ignore_accents
    )

    tiles: list[Tile] = [Tile.BLACK] * len(guess)
    remaining: Counter[str] = Counter()

    for i, (g, a) in enumerate(zip(normalized_guess, normalized_answer, strict=True)):
        if g.isspace():
            tiles[i] = Tile.SPACE
            if not a.isspace():
                remaining[a] += 1
            continue
        if g == a:
            tiles[i] = Tile.GREEN
        elif not a.isspace():
            remaining[a] += 1

    for i, (g, tile) in enumerate(zip(normalized_guess, tiles, strict=True)):
        if tile is not Tile.BLACK:
            continue
        if remaining[g] > 0:
            tiles[i] = Tile.YELLOW
            remaining[g] -= 1

    return tiles


def render(tiles: list[Tile]) -> str:
    return "".join(tile.value for tile in tiles)

File: web/src/scorer.ts
export type Tile = "green" | "yellow" | "black" | "space" | "empty";

export const TILE_EMOJI: Record<Exclude<Tile, "empty">, string> = {
  green: "🟩",
  yellow: "🟨",
  black: "⬛",
  space: "  ",
};

export type ScoreOptions = {
  caseInsensitive?: boolean;
  ignoreAccents?: boolean;
};

function stripAccents(text: string): string {
  return text.normalize("NFD").replaceAll(/\p{Mn}/gu, "");
}

function normalize(text: string, opts: Required<ScoreOptions>): string {
  let out = text;
  if (opts.ignoreAccents) out = stripAccents(out);
  if (opts.caseInsensitive) out = out.toLocaleLowerCase();
  return out;
}

/**
 * Wordle two-pass scoring. Greens consume their answer-letter first so a
 * duplicated letter in the guess can't double-claim a single occurrence
 * in the answer. Space positions are tagged separately to preserve word
 * boundaries visually without distorting the letter accounting.
 */
export function score(guess: string, answer: string, options: ScoreOptions = {}): Tile[] {
  const opts: Required<ScoreOptions> = {
    caseInsensitive: options.caseInsensitive ?? true,
    ignoreAccents: options.ignoreAccents ?? true,
  };

  const len = Math.max(guess.length, answer.length);
  if (len === 0) return [];

  const ng = normalize(guess, opts);
  const na = normalize(answer, opts);
  const tiles: Tile[] = Array.from({ length: len }, () => "empty" as Tile);
  const remaining = new Map<string, number>();

  for (let i = 0; i < len; i++) {
    const g = ng[i];
    const a = na[i];
    if (g === undefined) {
      tiles[i] = "empty";
      if (a !== undefined && !/\s/.test(a)) {
        remaining.set(a, (remaining.get(a) ?? 0) + 1);
      }
      continue;
    }
    if (/\s/.test(g)) {
      tiles[i] = "space";
      if (a !== undefined && !/\s/.test(a)) {
        remaining.set(a, (remaining.get(a) ?? 0) + 1);
      }
      continue;
    }
    if (a !== undefined && g === a) {
      tiles[i] = "green";
    } else {
      tiles[i] = "black";
      if (a !== undefined && !/\s/.test(a)) {
        remaining.set(a, (remaining.get(a) ?? 0) + 1);
      }
    }
  }

  for (let i = 0; i < len; i++) {
    if (tiles[i] !== "black") continue;
    const g = ng[i];
    if (g === undefined) continue;
    const left = remaining.get(g) ?? 0;
    if (left > 0) {
      tiles[i] = "yellow";
      remaining.set(g, left - 1);
    }
  }

  return tiles;
}

export function render(tiles: Tile[]): string {
  return tiles.map((t) => (t === "empty" ? "" : TILE_EMOJI[t])).join("");
}

export function isComplete(tiles: Tile[]): boolean {
  return tiles.length > 0 && tiles.every((t) => t === "green" || t === "space");
}
